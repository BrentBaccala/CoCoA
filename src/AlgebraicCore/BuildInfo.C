//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
