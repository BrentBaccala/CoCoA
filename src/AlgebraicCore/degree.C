//   Copyright (c)  2002-2009,2011  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/degree.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/assert.H"
#include "CoCoA/convert.H"
#include "CoCoA/error.H"
#include "CoCoA/utils.H"

#include <iostream>
using std::ostream;
#include<algorithm>
using std::max;
//#include <vector>
using std::vector;

namespace CoCoA
{

  inline void degree::CheckCompatible(const degree& d1, const degree& d2, const char* fn)
  {
    if (len(d1.myCoords) != len(d2.myCoords))
      CoCoA_ERROR(ERR::MixedDegrees, fn);
  }


  const BigInt& degree::operator[](long index) const
  {
    if (index < 0 || index >= len(myCoords))
      CoCoA_ERROR(ERR::BadDegIndex, "degree::operator[]");
    return myCoords[index];
  }


  void degree::mySetComponent(long index, const BigInt& N)
  {
    CoCoA_ASSERT(0 <= index && index < len(myCoords));
    myCoords[index] = N;
  }


  void degree::mySetComponent(long index, const MachineInt& n)
  {
    CoCoA_ASSERT(0 <= index && index < len(myCoords));
    myCoords[index] = BigInt(n);
  }


  degree& degree::operator+=(const degree& d)
  {
    degree::CheckCompatible(*this, d, "degree += degree");
    const long dim = len(myCoords);
    for (long i=0; i < dim; ++i)
      myCoords[i] += d[i];
    return *this;
  }


  degree& degree::operator-=(const degree& d)
  {
    degree::CheckCompatible(*this, d, "degree -= degree");
    const long dim = len(myCoords);
    for (long i=0; i < dim; ++i)
      myCoords[i] -= d[i];
    return *this;
  }


  bool IsZero(const degree& d)
  {
    const long dim = GradingDim(d);
    //??? return find_if(coords.begin(), coords.end(), NonZero) != coords.end();
    for (long i=0; i < dim; ++i)
      if (d[i] != 0) return false;
    return true;
  }


  degree operator+(const degree& d1, const degree& d2)
  {
    degree::CheckCompatible(d1, d2, "degree + degree");
    const long dim = GradingDim(d1);
    degree ans(dim);
    for (long i=0; i < dim; ++i)
      ans.mySetComponent(i, d1[i] + d2[i]);
    return ans;
  }


  degree operator-(const degree& d1, const degree& d2)
  {
    degree::CheckCompatible(d1, d2, "degree - degree");
    const long dim = GradingDim(d1);
    degree ans(dim);
    for (long i=0; i < dim; ++i)
      ans.mySetComponent(i, d1[i] - d2[i]);
    return ans;
  }


  degree top(const degree& d1, const degree& d2)
  {
    degree::CheckCompatible(d1, d2, "top");
    const long dim = GradingDim(d1);
    degree ans(dim);
    for (long i=0; i < dim; ++i)
      ans.mySetComponent(i, max(d1[i], d2[i]));
    return ans;
  }


  int cmp(const degree& d1, const degree& d2)
  {
    degree::CheckCompatible(d1, d2, "cmp(degree, degree)");
    // The rest is the same as the body of FastCmp; writing it explicitly
    // here avoids compiler foibles (i.e. choosing not to make FastCmp inline).
//     const long dim = GradingDim(d1);
    return FastCmp(d1, d2);
//     return LexCmp3(&d1.myCoords[0], &d1.myCoords[dim],
//                    &d2.myCoords[0], &d2.myCoords[dim]);
//     for (long i=0; i < dim; ++i)
//       if (d1[i] != d2[i])
//         return (d1[i] > d2[i] ? 1 : -1);
//     return 0;
  }


  ostream& operator<<(ostream& out, const degree& d)
  {
    const long dim = GradingDim(d);
    if (dim == 0) return out << "()";     // no grading -- graded over N^0
    //  if (dim == 1) return out << d[0]; // omit parens if grading is over N
    // General case
    out << "(";
    for (long i=0; i < dim-1; ++i)
      out << d[i] << ", ";
    out << d[dim-1] << ")";
    return out;
  }

  void SetComponent(degree& d, long index, const BigInt& N)
  {
    if (index < 0 || index >= GradingDim(d))
      CoCoA_ERROR(ERR::BadDegIndex, "SetComponent(degree, index, N)");
    d.mySetComponent(index, N);
  }

  void SetComponent(degree& d, long index, const MachineInt& n)
  {
    if (index < 0 || index >= GradingDim(d))
      CoCoA_ERROR(ERR::BadDegIndex, "SetComponent(degree, index, n)");
    d.mySetComponent(index, n);
  }


} // end of namespace CoCoA

// RCS header/log
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/degree.C,v 1.11 2014/07/14 15:09:34 abbott Exp $
// $Log: degree.C,v $
// Revision 1.11  2014/07/14 15:09:34  abbott
// Summary: Added include of utils.H
// Author: JAA
//
// Revision 1.10  2012/05/28 09:18:20  abbott
// Created IntOperations which gathers together all operations on
// integers (both big and small).  Many consequential changes.
//
// Revision 1.9  2011/11/09 14:29:37  bigatti
// -- renamed MachineInteger --> MachineInt
//
// Revision 1.8  2011/08/14 15:52:16  abbott
// Changed ZZ into BigInt (phase 1: just the library sources).
//
// Revision 1.7  2011/03/10 16:39:33  abbott
// Replaced (very many) size_t by long in function interfaces (for rings,
// PPMonoids and modules).  Also replaced most size_t inside fn defns.
//
// Revision 1.6  2009/02/20 09:55:43  bigatti
// -- changed: now also degrees with GradingDim=1 are printed with parentheses
//
// Revision 1.5  2008/12/17 11:53:08  abbott
// Indexes into degree objects are now MachineInts rather than size_t.
// Hope this proves not to be a duff idea.
//
// Revision 1.4  2008/12/16 21:36:41  abbott
// Changed long into MachineInt for specifying values when setting components.
//
// Revision 1.3  2008/04/21 12:32:54  abbott
// Corrected size_t into std::size_t in several header files; in some cases,
// replaced size_t with MachineInt (with consequent changes to impl files).
//
// Revision 1.2  2007/10/30 17:14:06  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.3  2006/10/06 14:04:14  cocoa
// Corrected position of #ifndef in header files.
// Separated CoCoA_ASSERT into assert.H from config.H;
// many minor consequential changes (have to #include assert.H).
// A little tidying of #include directives (esp. in Max's code).
//
// Revision 1.2  2006/08/07 21:23:25  cocoa
// Removed almost all publicly visible references to SmallExponent_t;
// changed to long in all PPMonoid functions and SparsePolyRing functions.
// DivMask remains to sorted out.
//
// Revision 1.1.1.1  2006/05/30 11:39:37  cocoa
// Imported files
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.3  2005/08/08 16:36:32  cocoa
// Just checking in before going on holiday.
// Don't really recall what changes have been made.
// Added IsIndet function for RingElem, PPMonoidElem,
// and a member function of OrdvArith.
// Improved the way failed assertions are handled.
//
// Revision 1.2  2005/06/27 16:23:04  cocoa
// -- Added: GradingDim, operator+=, operator-=
// -- Removed: IsSubtractable (and calls to it in operator-)
//
// Revision 1.1.1.1  2005/05/03 15:47:31  cocoa
// Imported files
//
// Revision 1.5  2005/04/29 14:34:14  cocoa
// -- changed: in mySetComponent  myCoords[pos] = ZZ(value);  (was "= value")
//
// Revision 1.4  2005/04/29 14:21:17  cocoa
// -- changed: ElementType is now ZZ (was int)
//
// Revision 1.3  2005/04/20 15:40:47  cocoa
// Major change: modified the standard way errors are to be signalled
// (now via a macro which records filename and line number).  Updated
// documentation in error.txt accordingly.
//
// Improved the documentation in matrix.txt (still more work to be done).
//
// Revision 1.2  2005/04/19 14:06:03  cocoa
// Added GPL and GFDL licence stuff.
//
// Revision 1.1.1.1  2005/01/27 15:12:13  cocoa
// Imported files
//
// Revision 1.5  2004/11/12 15:49:29  cocoa
// Tidying prior to 0.90 release.
// (a) documentation improved (or marked as poor)
// (b) sundry minor improvements to the code
//
// Revision 1.4  2004/11/11 14:08:19  cocoa
// -- minor changes for doxygen
// -- moved CVS log to the bottom
//
// Revision 1.3  2004/05/27 16:14:02  cocoa
// Minor revision for new coding conventions.
//
// Revision 1.2  2004/01/28 15:29:42  cocoa
// Added IsZero function, and changed name of coordinate type used
// in the representation of degree objects.
//
// Revision 1.1.1.1  2003/09/24 12:55:43  cocoa
// Imported files
//
// Revision 1.3  2003/06/23 16:23:03  abbott
// Minor cleaning prior to public release.
//
// Revision 1.2  2003/05/14 17:09:08  abbott
// Added definitions of operator[] (read-only access),
// top and SetComponent.
//
// Revision 1.1  2002/12/18 18:19:20  abbott
// Initial revision
//
