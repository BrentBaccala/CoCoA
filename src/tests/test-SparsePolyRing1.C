//   Copyright (c)  2007  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/matrix.H"
#include "CoCoA/MatrixForOrdering.H"
#include "CoCoA/MatrixView.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDistrMPolyClean.H"
#include "CoCoA/RingDistrMPolyInlFpPP.H"
#include "CoCoA/RingDistrMPolyInlPP.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/SparsePolyRing.H"
#include "CoCoA/symbol.H"
#include "CoCoA/degree.H"

using namespace CoCoA;

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for SparsePolyRing functions on RingElem
// functions: indets, *, wdeg, StdDeg, IsHomog, homog, gcd, CmpWDeg
// environments: DMP (Fp, ZZ), DMPI (Q), DMPII
//----------------------------------------------------------------------

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)


void TestSparsePolyRing(SparsePolyRing P)
{
  cout << "TESTING: " << P << endl << endl;
  cout << std::boolalpha; // prints true/false for bool

  
  const vector<RingElem>& x = indets(P);
  RingElem f1 = x[0]*x[0] + 3*x[1] -2,
    f2 = x[1]*x[1] - 5*x[0],
    f3 = x[0]-x[1];
  ConstRefRingElem h = x[NumIndets(P)-1];

  cout << "Given f1 = " << f1 << endl;
  cout << "  and f2 = " << f2 << endl << endl;
 
  cout << "f1*f2   gives  " << f1*f2 << endl;
  //  cout << "gcd(f1,f2)   gives  " << gcd(f1,f2) << endl;
  cout << "wdeg(f1)   gives  " << wdeg(f1) << endl;
  cout << "wdeg(f2)   gives  " << wdeg(f2) << endl;
  cout << "StdDeg(f1)   gives  " << StdDeg(f1) << endl;
  cout << "StdDeg(f2)   gives  " << StdDeg(f2) << endl;
  cout << "IsHomog(f1)   gives  " << IsHomog(f1) << endl;
  cout << "IsHomog(f2)   gives  " << IsHomog(f2) << endl;
  cout << "IsHomog(f3)   gives  " << IsHomog(f3) << endl;
  cout << "CmpWDeg(f1,f2)        gives  " << sign(CmpWDeg(f1,f2)) << endl; // CmpWDeg guarantees only the sign of result
  if (GradingDim(P)>0)
  {
    cout << "LF(f3)   gives  " << LF(f3) << endl;
    cout << "CmpWDegPartial(f1,f2) gives  "
         << sign(CmpWDegPartial(f1,f2,1)) << endl;  // CmpWDeg guarantees only the sign of result
    cout << "IsHomogPartial(f3,1)  gives  "
                   << IsHomogPartial(f3,1) << endl;
  }
  if (GradingDim(P)<2)
  {
    cout << "  -- homogenizing with h = " << h << endl;
    cout << "homog(f1, h)   gives  " << homog(f1,h) << endl;
    cout << "homog(f2, h)   gives  " << homog(f2,h) << endl;
    cout << "homog(f3, h)   gives  " << homog(f3,h) << endl;
  }

  if ( IsField(CoeffRing(P)) || IsTrueGCDDomain(CoeffRing(P)))
  {
    TEST_ASSERT( IsInvertible(gcd(f1*f2, f1*f3)/f1) );
    TEST_ASSERT( IsInvertible(gcd(f1*f2, f2*f3)/f2) );
    TEST_ASSERT( IsInvertible(gcd(f1*f3, f2*f3)/f3) );
  }
  if ( IsFractionFieldOfGCDDomain(CoeffRing(P)) )
  {
    TEST_ASSERT( CommonDenom(x[0]/3+x[1]/2) ==  6);
    TEST_ASSERT( ClearDenom(x[0]/3+x[1]/2) ==  2*x[0]+3*x[1]);
  }
  
  // Check some computations (without printing any results).
  cout << "one(P)  " << one(P) << endl;
  TEST_ASSERT(IsOne(one(P)));
  TEST_ASSERT(!IsOne(f1));
  TEST_ASSERT(!IsOne(f2));

  RingElem f1f2 = f1*f2;
  RingElem f2f1 = f2*f1;
  TEST_ASSERT(f1f2 == f2f1);
  TEST_ASSERT(f1 != f2);
  TEST_ASSERT(f1f2 != f1);
  TEST_ASSERT(f1f2 != f2);
  TEST_ASSERT(IsDivisible(f1f2, f1));
  TEST_ASSERT(IsDivisible(f1f2, f2));
  TEST_ASSERT(!IsDivisible(f1, f2));
  TEST_ASSERT(!IsDivisible(f2, f1));
  TEST_ASSERT(f1f2/f1 == f2);
  TEST_ASSERT(f1f2/f2 == f1);
  TEST_ASSERT(power(f1,2) == f1*f1);
  TEST_ASSERT(power(f2,2) == f2*f2);
  TEST_ASSERT((power(f1f2,2))/(f1f2*f1f2) == 1);
  TEST_ASSERT(deriv(f1f2, x[0]) == deriv(f1, x[0])*f2 + f1*deriv(f2, x[0]));
  TEST_ASSERT(deriv(x[1]+2*x[0], x[0]) == 2);

  //  TEST_ASSERT(gcd(f1*f1f2, f2*f1f2) == f1f2);

  TEST_ASSERT(CmpWDeg(f1, one(P)) > 0);
  TEST_ASSERT(CmpWDeg(f2, one(P)) > 0);
  TEST_ASSERT(CmpWDeg(one(P), f1) < 0);
  TEST_ASSERT(CmpWDeg(one(P), f2) < 0);
  TEST_ASSERT(CmpWDeg(f1f2, f1) > 0);
  TEST_ASSERT(CmpWDeg(f1f2, f2) > 0);
  TEST_ASSERT(CmpWDeg(f1, f1f2) < 0);
  TEST_ASSERT(CmpWDeg(f2, f1f2) < 0);

  TEST_ASSERT(CmpWDeg(power(f1f2, 3), f1f2) > 0);

  f2f1 = 1;
  TEST_ASSERT(IsOne(f2f1));

  TEST_ASSERT(wdeg(f1) + wdeg(one(P)) == wdeg(f1));
  TEST_ASSERT(wdeg(f2) + wdeg(one(P)) == wdeg(f2));
  TEST_ASSERT(wdeg(f1) + wdeg(f2) == wdeg(f1f2));

  vector<long> exps0(NumIndets(P));  exps0[0] = 1;
  vector<long> exps1(NumIndets(P));  exps1[1] = 1;
  RingElem g0(x[0]), g1(x[1]);
  PushBack(g0, -one(CoeffRing(P)), exps1);
  PushFront(g1, one(CoeffRing(P)), exps0);
  TEST_ASSERT(g0 == x[0]-x[1]);
  TEST_ASSERT(g1 == x[0]+x[1]);

  cout << "------------------------------------------------" << endl << endl;
}


void program()
{
  GlobalManager CoCoAFoundations(UseSymmResidues);
  const vector<RingElem> v(3, one(RingZZ()));
  ConstMatrixView I = IdentityMat(RingZZ(), 3);
  PPOrdering Deg2(NewMatrixOrdering(3, 2, NewMatCompleteOrd(ConcatVer(RowMat(v), I))));

  const QuotientRing Fp = NewZZmod(101);
  const SparsePolyRing ZZxy = NewPolyRing(RingZZ(), 2);
  const SparsePolyRing Fpy = NewPolyRing_DMPI(Fp, 3);
  const SparsePolyRing FpyII = NewPolyRing_DMPII(Fp, 4);
  const SparsePolyRing Qx = NewPolyRing_DMPI(RingQQ(), 4);
  const SparsePolyRing Q2x = NewPolyRing_DMPI(RingQQ(), symbols("x","y","z"), Deg2);

  TestSparsePolyRing(ZZxy);
  TestSparsePolyRing(Fpy);
  TestSparsePolyRing(FpyII);
  TestSparsePolyRing(Qx);
  TestSparsePolyRing(Q2x);
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
