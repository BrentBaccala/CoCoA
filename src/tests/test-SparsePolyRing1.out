TESTING: RingWithID(3,"ZZ[x[0],x[1]]")

Given f1 = x[0]^2 +3*x[1] -2
  and f2 = x[1]^2 -5*x[0]

f1*f2   gives  x[0]^2*x[1]^2 -5*x[0]^3 +3*x[1]^3 -15*x[0]*x[1] -2*x[1]^2 +10*x[0]
wdeg(f1)   gives  (2)
wdeg(f2)   gives  (2)
StdDeg(f1)   gives  2
StdDeg(f2)   gives  2
IsHomog(f1)   gives  false
IsHomog(f2)   gives  false
IsHomog(f3)   gives  true
CmpWDeg(f1,f2)        gives  0
LF(f3)   gives  x[0] -x[1]
CmpWDegPartial(f1,f2) gives  0
IsHomogPartial(f3,1)  gives  true
  -- homogenizing with h = x[1]
homog(f1, h)   gives  x[0]^2 +x[1]^2
homog(f2, h)   gives  -5*x[0]*x[1] +x[1]^2
homog(f3, h)   gives  x[0] -x[1]
one(P)  1
------------------------------------------------

TESTING: RingWithID(4,"RingWithID(2)[x[0],x[1],x[2]]")

Given f1 = x[0]^2 +3*x[1] -2
  and f2 = x[1]^2 -5*x[0]

f1*f2   gives  x[0]^2*x[1]^2 -5*x[0]^3 +3*x[1]^3 -15*x[0]*x[1] -2*x[1]^2 +10*x[0]
wdeg(f1)   gives  (2)
wdeg(f2)   gives  (2)
StdDeg(f1)   gives  2
StdDeg(f2)   gives  2
IsHomog(f1)   gives  false
IsHomog(f2)   gives  false
IsHomog(f3)   gives  true
CmpWDeg(f1,f2)        gives  0
LF(f3)   gives  x[0] -x[1]
CmpWDegPartial(f1,f2) gives  0
IsHomogPartial(f3,1)  gives  true
  -- homogenizing with h = x[2]
homog(f1, h)   gives  x[0]^2 +3*x[1]*x[2] -2*x[2]^2
homog(f2, h)   gives  x[1]^2 -5*x[0]*x[2]
homog(f3, h)   gives  x[0] -x[1]
one(P)  1
------------------------------------------------

TESTING: RingWithID(5,"RingWithID(2)[x[0],x[1],x[2],x[3]]")

Given f1 = x[0]^2 +3*x[1] -2
  and f2 = x[1]^2 -5*x[0]

f1*f2   gives  x[0]^2*x[1]^2 -5*x[0]^3 +3*x[1]^3 -15*x[0]*x[1] -2*x[1]^2 +10*x[0]
wdeg(f1)   gives  (2)
wdeg(f2)   gives  (2)
StdDeg(f1)   gives  2
StdDeg(f2)   gives  2
IsHomog(f1)   gives  false
IsHomog(f2)   gives  false
IsHomog(f3)   gives  true
CmpWDeg(f1,f2)        gives  0
LF(f3)   gives  x[0] -x[1]
CmpWDegPartial(f1,f2) gives  0
IsHomogPartial(f3,1)  gives  true
  -- homogenizing with h = x[3]
homog(f1, h)   gives  x[0]^2 +3*x[1]*x[3] -2*x[3]^2
homog(f2, h)   gives  x[1]^2 -5*x[0]*x[3]
homog(f3, h)   gives  x[0] -x[1]
one(P)  1
------------------------------------------------

TESTING: RingWithID(6,"QQ[x[0],x[1],x[2],x[3]]")

Given f1 = x[0]^2 +3*x[1] -2
  and f2 = x[1]^2 -5*x[0]

f1*f2   gives  x[0]^2*x[1]^2 -5*x[0]^3 +3*x[1]^3 -15*x[0]*x[1] -2*x[1]^2 +10*x[0]
wdeg(f1)   gives  (2)
wdeg(f2)   gives  (2)
StdDeg(f1)   gives  2
StdDeg(f2)   gives  2
IsHomog(f1)   gives  false
IsHomog(f2)   gives  false
IsHomog(f3)   gives  true
CmpWDeg(f1,f2)        gives  0
LF(f3)   gives  x[0] -x[1]
CmpWDegPartial(f1,f2) gives  0
IsHomogPartial(f3,1)  gives  true
  -- homogenizing with h = x[3]
homog(f1, h)   gives  x[0]^2 +3*x[1]*x[3] -2*x[3]^2
homog(f2, h)   gives  x[1]^2 -5*x[0]*x[3]
homog(f3, h)   gives  x[0] -x[1]
one(P)  1
------------------------------------------------

TESTING: RingWithID(8,"QQ[x,y,z]")

Given f1 = x^2 +3*y -2
  and f2 = y^2 -5*x

f1*f2   gives  x^2*y^2 -5*x^3 +3*y^3 -15*x*y -2*y^2 +10*x
wdeg(f1)   gives  (2, 2)
wdeg(f2)   gives  (2, 0)
StdDeg(f1)   gives  2
StdDeg(f2)   gives  2
IsHomog(f1)   gives  false
IsHomog(f2)   gives  false
IsHomog(f3)   gives  false
CmpWDeg(f1,f2)        gives  1
LF(f3)   gives  x
CmpWDegPartial(f1,f2) gives  0
IsHomogPartial(f3,1)  gives  true
one(P)  1
------------------------------------------------

